src/ir/from_ast/astconv.rs

```diff
@@ -860,94 +850,79 @@ impl<'ctx, 'prog> BuildCtx<'ctx, 'prog> {
+            // enable source information saving if this is main or an external.
+            if body.is_none() || Some(idx) == ctx.entrypoint {
+                builder.comp.src_info = Some(InterfaceSrc::new(sig.name.copy()))
+            }
+            // add the file to the externals map if it exists
+            if let Some(file) = file {
+                ctx.externals.entry(file).or_default().push(idx);
+            }
 
-    let mut ctx = ir::Context::default();
-    for (file, exts) in ns.externs {
-        for ext in exts {
-            let idx = sig_map.get(&ext.name).unwrap().idx;
-            log::debug!("Converting external {}: {}", ext.name, idx);
-            let ir_ext = BuildCtx::external(&ctx, ext);
-            ctx.comps.checked_add(idx, ir_ext);
-            ctx.externals.entry(file.clone()).or_default().push(idx);
-        }
-    }
+            // compile the signature
+            builder.comp.cmds = builder.sig(&sig);
 
-    // TODO: Need to handle recursive components as well as mutually recursive components by adding signature ports first in declaration, so that foreign keys can be resolved
-    // declare all components in the proper order
-    for (cidx, comp) in ns.components.iter().enumerate() {
-        let idx = sig_map.get(&comp.sig.name).unwrap().idx;
+            let irsig = Sig::new(idx, &builder.comp, &sig);
+            ((idx, builder, body), (sig.name.take(), irsig))
+        })
+        .unzip();
 
-        let mut ir_comp = ir::Component::new(false);
-        if Some(cidx) == main_idx {
-            ctx.entrypoint = Some(idx);
-            ir_comp.src_info =
-                Some(ir::InterfaceSrc::new(comp.sig.name.copy()));
-        }
-        ctx.comps.checked_add(idx, ir_comp);
-    }
+    // Add the signature map to each builder
+    builders.iter_mut().for_each(|(_, builder, _)| {
+        builder.sigs = Some(&sig_map);
+    });
 
-    // create a dummy component to be swapped into the context
-    let mut curr_comp = ir::Component::new(false);
-    for cidx in order {
-        let comp = std::mem::take(&mut ns.components[cidx]);
-        log::debug!("Compiling component {}", comp.sig.name);
-        let idx = sig_map.get(&comp.sig.name).unwrap().idx;
-
-        // Needs to swap here because we have to own the current component while leaving the context immutable.
-        // TODO: Find a solution here that will still allow us to properly generate Foreign keys in recursive components.
-        std::mem::swap(ctx.get_mut(idx), &mut curr_comp);
-        BuildCtx::comp(&ctx, &mut curr_comp, comp, &sig_map);
-        // swap the component back into place
-        std::mem::swap(ctx.get_mut(idx), &mut curr_comp);
-    }
+    // Compiles and adds all the commands here
+    // Need to do this before adding all the components because each builder borrows the context immutably
+    builders.into_iter().for_each(|(idx, mut builder, cmds)| {
```

Comment:
Can this be a `for` loop instead? Usually, if the iterator chain is just `into_iter().for_each(...)`, I prefer using `for` loops